Bug: Made it so that 0 was included in uint8

verhalenn opened the pull request:
- [ ] closes #14401 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master | flake8 --diff`
- [ ] whatsnew entry

Decided to restart. Sorry for the inconvenience. -> #14472

Commit:
Made it so that 0 was included in uint8

Commit:
Added a test to check uint8 with 0

Commit:
Added release note to issue 14401 resolve.

sinhrks commented:
Thx for the PR. pls rebase existing PR from next.

sinhrks commented on doc/source/whatsnew/v0.19.1.txt:
```diff
@@ -45,3 +45,4 @@ Bug Fixes
 
 - Bug in ``pd.concat`` where names of the ``keys`` were not propagated to the resulting ``MultiIndex`` (:issue:`14252`)
 - Bug in ``MultiIndex.set_levels`` where illegal level values were still set after raising an error (:issue:`13754`)
+- Bug in ``pd.to_numeric`` where it would not downcast a 0 to a uint8 (:issue:`14404`)
```
pls include #14401.

jreback commented on pandas/tools/tests/test_util.py:
```diff
@@ -391,6 +391,14 @@ def test_downcast(self):
 
+        #check that 0 works as a unsigned downcast
+
+        data = [0, 1, 2, 3]
```
check this for all of the downcast parameters (where it *should* be included)

verhalenn replied:
Did you want me to just edit everything between line 327 and line 368 in that test to include 0's in the arrays? Trying it before and after, it fails before and passes after. That being said if you don't want me to touch it for some reason it's relatively easy to make a new one. Also is there a reason that 340 and 341 is redundant?

Commit:
Edited mistakes in whatsnew

Commit:
Added tests for the max and min values of all dtypes to to_numeric

verhalenn commented:
Added a series of tests which applied a pandas series that included the minimum and maximum values of each dtype to a to_numeric downcast to see if it would downcast to the appropriate dtype. It did not add to uint64 but I'm not sure if there were plans to include uint64 so I'll open up a new issue. Also edited the whatsnew.

jreback commented on doc/source/whatsnew/v0.19.1.txt:
```diff
@@ -45,3 +45,4 @@ Bug Fixes
 
 - Bug in ``pd.concat`` where names of the ``keys`` were not propagated to the resulting ``MultiIndex`` (:issue:`14252`)
 - Bug in ``MultiIndex.set_levels`` where illegal level values were still set after raising an error (:issue:`13754`)
+- Bug in ``pd.to_numeric`` where it would not downcast a 0 properly. (:issue:`14401`)
```
say ``downcast='unsigned'`` as a kwargs is passed

jreback commented on pandas/tools/tests/test_util.py:
```diff
@@ -401,6 +401,32 @@ def test_downcast(self):
             tm.assert_numpy_array_equal(res, expected)
 
+        # check that the smallest and largest values in each integer type pass to each type.
+        int8 = [-128, 127]
```
don't do this explicty, use ``np.iinfo(dtype).min/max`` as these vary by platform

jreback replied:
actually on 2nd though, might be ok if appveyor passes

jreback replied:
can you do this in a loop instead (eg. a loop of tuples of the start,stop, dtype)

verhalenn replied:
I was actually thinking dictionary if that's okay. And I believe that the numpy types are standard across platforms but I'm not positive which is why I didn't bother with the np.info(dtype).min/max but for safeties sake I'll include it.

Commit:
Changed the tests so that it iterated through a dictionary.

I also added a series of other tests that included making sure that a
dtype shifted dtype once it reached the next value and once again
edited the whatsnew.

Commit:
Changed the test to work with python 3.x

jreback commented on doc/source/whatsnew/v0.19.1.txt:
```diff
@@ -48,3 +48,6 @@ Bug Fixes
 - Bug in ``MultiIndex.set_levels`` where illegal level values were still set after raising an error (:issue:`13754`)
 - Bug in ``DataFrame.to_json`` where ``lines=True`` and a value contained a ``}`` character (:issue:`14391`)
 - Bug in ``df.groupby`` causing an ``AttributeError`` when grouping a single index frame by a column and the index level (:issue`14327`)
+- Bug in ``pd.to_numeric`` where it would not downcast a 0 to a uint8 (:issue:`14404`)
```
just make a single entry

jreback commented on pandas/tools/tests/test_util.py:
```diff
@@ -401,6 +401,62 @@ def test_downcast(self):
             res = pd.to_numeric(data, downcast=downcast)
             tm.assert_numpy_array_equal(res, expected)
 
+        # check that the smallest and largest values in each integer type pass to each type.
```
this is still overly verbose. do something like.

```
checks = [('int8', 'integer', [np.iinfo(np.int8).min, np.iinfo(np.int8).max]), 
                  .....
               ]

for dtype, downcast, min_max in checks:
     ....
```
